Improve error message as "x is not a registered account" from "Unknown account".

deepakhb2 opened the pull request:
### Description

This PR is to improve the error message to be more meaningful.

### Other change

Accounts contract inherits InitializableV2.

### Tested



### Related issues

- Fixes #7173 

### Backwards compatibility

This change is backward compatible.

deepakhb2 committed:
Improve error message as "x is not a registered account".

deepakhb2 committed:
Inherit from InitializableV2 Contract.

m-chrzan commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -129,7 +129,10 @@ contract Accounts is
-    require(isAccount(msg.sender), "Unknown account");
+    require(
+      isAccount(msg.sender),
+      string(abi.encodePacked(msg.sender, " is not registered account"))
```
Typo: "is not *a* registered account". Here and on other fixed lines.

m-chrzan reviewed (commented):
Hey, thanks for this! The CI check is failing with "Accounts has empty bytecode", I think this is due to inheriting from the new InitializableV2 - the Accounts contract now needs to call InitializableV2's constructor. You should be fine just copying the constructor and doc comment from [ReleaseGold](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/ReleaseGold.sol#L163).

Once that's done, you'll also need to pass `true` to all Accounts deployments in unit tests. Running `yarn build` in the protocol package should give you a list of all spots where this needs to be adjusted.

You should be able to fix your #7517 PR in a similar way. Let me know if this is enough context or you have any questions!

nambrot commented:
I don't know the gas overhead of concatenating the address into the string, but are we sure it is negligible?

deepakhb2 committed:
Fix the typo in error message.
Update tests to fix fails due to InitializableV2 change for Accounts.

deepakhb2 commented:
> I don't know the gas overhead of concatenating the address into the string, but are we sure it is negligible?

Good point, I think it shouldn't consume more gas for this inline code. I will use Remix to find the exact number.

deepakhb2 committed:
Update patch contract version for Accounts contract.

deepakhb2 commented:
@nambrot It costs 1113 gas on remix and supporting blog in medium explaining gas consumption [link](https://medium.com/coinmonks/gas-cost-of-solidity-library-functions-dbe0cedd4678). I hope this helps and I think it is negligible. I am not sure as I am new to this tech stack.

AlexBHarley reviewed (commented):
Nice @deepakhb2! Another great UX improvement

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -129,7 +135,10 @@ contract Accounts is
+    require(
+      isAccount(msg.sender),
+      string(abi.encodePacked(msg.sender, " is not a registered account"))
+    );
```
should we make this a modifier to DRY up the code a bit?
https://docs.soliditylang.org/en/v0.6.2/contracts.html#function-modifiers
not sure how the compiler optimizes but the string might also be redundant in the bytecode

deepakhb2 replied:
@yorhodes thank you for pointing this out. I will make this change.

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -148,7 +157,10 @@ contract Accounts is
-    require(isAccount(msg.sender), "Unknown account");
+    require(
+      isAccount(msg.sender),
+      string(abi.encodePacked(msg.sender, " is not a registered account"))
```
It's not as useful as it seems to include the `msg.sender` in the revert reason
In order to get the revert reason, clients use `eth_call` RPC which takes as input the encoded transaction
This encoded transaction already contains enough information to derive the `sender`

yorhodes replied:
cc @nambrot @deepakhb2
Any gas overhead isn't worth it imo

asaj replied:
I don't think we care about consuming ~1k extra gas when the tx reverts, but would prefer not to consume that gas if the tx succeeds. Does the string concatenation actually cost any gas if the condition is true? I suspect it compiles the same as:
```
if (!isAccount(msg.sender)) {
  revert(string(abi.encodePacked(msg.sender, " is not a registered account")))
}
```

yorhodes replied:
my recommendation is to make the revert reason static and avoid the question altogether
don't see much downside
```suggestion
      "sender is not a registered account"
```

nambrot replied:
That was @m-chrzan 's theory as well

deepakhb2 replied:
@yorhodes Initially, I thought of it but I didn't raise as I wasn't sure about actual use case of this message. I think, sender will be know the address used for the transaction and It makes sense to respond with 'Sender is not a registered account'. I will make this change as well if this is the final call.

yorhodes reviewed (changes requested):
A couple comments

deepakhb2 committed:
Merge branch 'master' of github.com:celo-org/celo-monorepo into issue/7173-error-message

deepakhb2 committed:
Add modifier isSenderRegistered and use it in all other function.
Modifier check if sender is registered account.
Remove string concatenation.

deepakhb2 committed:
Add modifier method after public keyword for function.

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -85,6 +85,17 @@ contract Accounts is
+  constructor(bool test) public InitializableV2(test) {}
+
+  modifier isSenderRegistered {
+    require(isAccount(msg.sender), "Sender is not a registered account");
```
We often use the prefix `only` for descriptive modifier names
Additionally, we need to parameterize this because `msg.sender` isn't always the account we are checking

```suggestion
  modifier onlyRegisteredAccount(address account) {
    require(isAccount(account), "Account is not registered");
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -128,8 +139,7 @@ contract Accounts is
    */
-  function setName(string memory name) public {
-    require(isAccount(msg.sender), "Unknown account");
+  function setName(string memory name) public isSenderRegistered {
```
```suggestion
  function setName(string memory name) public onlyRegisteredAccount(msg.sender) {
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -147,8 +157,10 @@ contract Accounts is
-    require(isAccount(msg.sender), "Unknown account");
+  function setWalletAddress(address walletAddress, uint8 v, bytes32 r, bytes32 s)
+    public
+    isSenderRegistered
```
```suggestion
    onlyRegisteredAccount(msg.sender)
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -173,8 +185,7 @@ contract Accounts is
    */
-  function setMetadataURL(string calldata metadataURL) external {
-    require(isAccount(msg.sender), "Unknown account");
+  function setMetadataURL(string calldata metadataURL) external isSenderRegistered {
```
```suggestion
  function setMetadataURL(string calldata metadataURL) external onlyRegisteredAccount(msg.sender) {
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -401,8 +412,7 @@ contract Accounts is
    */
-  function getVoteSigner(address account) public view returns (address) {
-    require(isAccount(account), "Unknown account");
+  function getVoteSigner(address account) public view isSenderRegistered returns (address) {
```
```suggestion
  function getVoteSigner(address account) public view onlyRegisteredAccount(account) returns (address) {
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -412,8 +422,7 @@ contract Accounts is
    */
-  function getValidatorSigner(address account) public view returns (address) {
-    require(isAccount(account), "Unknown account");
+  function getValidatorSigner(address account) public view isSenderRegistered returns (address) {
```
```suggestion
  function getValidatorSigner(address account) public view onlyRegisteredAccount(account) returns (address) {
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -423,8 +432,7 @@ contract Accounts is
    */
-  function getAttestationSigner(address account) public view returns (address) {
-    require(isAccount(account), "Unknown account");
+  function getAttestationSigner(address account) public view isSenderRegistered returns (address) {
```
```suggestion
  function getAttestationSigner(address account) public view onlyRegisteredAccount(account) returns (address) {
```

yorhodes commented on packages/protocol/contracts/common/Accounts.sol:
```diff
@@ -574,8 +582,7 @@ contract Accounts is
    */
-  function authorize(address authorized, uint8 v, bytes32 r, bytes32 s) private {
-    require(isAccount(msg.sender), "Unknown account");
+  function authorize(address authorized, uint8 v, bytes32 r, bytes32 s) private isSenderRegistered {
```
```suggestion
  function authorize(address authorized, uint8 v, bytes32 r, bytes32 s) private onlyRegisteredAccount(msg.sender) {
```

yorhodes reviewed (changes requested):
be careful about `msg.sender` versus an `account` parameter

deepakhb2 committed:
Change modifier name from isSenderRegistered to onlyRegisteredAccount.
Pass the sender as parameter to modifier to make sure always correct account is checked.

deepakhb2 committed:
Update message.

deepakhb2 committed:
The parameters to modifier is account in few cases.

yorhodes reviewed (approved):
nice UX improvement!

deepakhb2 committed:
Merge branch 'master' of github.com:celo-org/celo-monorepo into issue/7173-error-message

yorhodes commented:
I am working on a resolution to the `test-release` issue. @deepakhb2 there is no further action needed from you at this time and I will make sure that your bounty assignment doesn't expire.

deepakhb2 commented:
@yorhodes This PR has conflicts. Is it still make sense to resolve the conflicts to see everything works as expected?
